mava/systems/sac/ff_isac.py

```diff
@@ -0,0 +1,600 @@
+        assert isinstance(obs, Observation)  # mypy thinks it's Observation | ObservationGlobalState
+
+        key, explore_key = jax.random.split(key)
+        action = jax.random.uniform(explore_key, full_action_shape)
+        next_obs, env_state, buffer_state, metrics = step(action, obs, env_state, buffer_state)
+
+        t += cfg.arch.num_envs
+        learner_state = carry._replace(
+            obs=next_obs, env_state=env_state, buffer_state=buffer_state, t=t, key=key
+        )
+        return learner_state, metrics
+
+    scanned_update = lambda state: lax.scan(update_epoch, state, None, length=cfg.system.epochs)
+    scanned_act = lambda state: lax.scan(act, state, None, length=cfg.system.rollout_length)
+
+    # Act loop -> sample -> update loop
+    def update_step(carry: LearnerState, _: Any) -> Tuple[LearnerState, Tuple[Metrics, Metrics]]:
+        """Act, sample, learn."""
+
+        obs, env_state, buffer_state, params, opt_states, t, key = carry
+        key, act_key, learn_key = jax.random.split(key, 3)
+        # Act
+        act_state = (params.actor, obs, env_state, buffer_state, act_key)
+        (_, next_obs, env_state, buffer_state, _), metrics = scanned_act(act_state)
+
+        # Sample and learn
+        learn_state = (buffer_state, params, opt_states, t, learn_key)
+        (buffer_state, params, opt_states, _, _), losses = scanned_update(learn_state)
+
+        t += cfg.arch.num_envs * cfg.system.rollout_length
+        return (
+            LearnerState(next_obs, env_state, buffer_state, params, opt_states, t, key),
+            (metrics, losses),
+        )
+
+    # pmap and scan over explore and update_step
+    # Make sure to not do num_envs explore steps (could fill up the buffer too much).
+    explore_steps = cfg.system.explore_steps // cfg.arch.num_envs
+    pmaped_explore = jax.pmap(
+        jax.vmap(
+            lambda state: lax.scan(explore, state, None, length=explore_steps),
+            axis_name="batch",
+        ),
+        axis_name="device",
+        donate_argnums=0,
+    )
+    pmaped_update_step = jax.pmap(
+        jax.vmap(
+            lambda state: lax.scan(update_step, state, None, length=cfg.system.scan_steps),
+            axis_name="batch",
+        ),
+        axis_name="device",
+        donate_argnums=0,
+    )
+
+    return pmaped_explore, pmaped_update_step
+
+
+def run_experiment(cfg: DictConfig) -> float:
+    # Add runtime variables to config
+    cfg.arch.n_devices = len(jax.devices())
+    cfg = check_total_timesteps(cfg)
+
+    # Number of env steps before evaluating/logging.
+    steps_per_rollout = int(cfg.system.total_timesteps // cfg.arch.num_evaluation)
+    # Multiplier for a single env/learn step in an anakin system
+    anakin_steps = cfg.arch.n_devices * cfg.system.update_batch_size
+    # Number of env steps in one anakin style update.
+    anakin_act_steps = anakin_steps * cfg.arch.num_envs * cfg.system.rollout_length
```

Comment:
Ah I quite liked this, what else would you call it. I went with anakin because it seems to encapsulate that notion of _steps over distributed devices_, maybe `distributed_steps`? Open to other names if you have any :smile:

Reply:
Ja, I don't have a better name off the cuff. Let's keep it as is for now. We can probably think about a better name when we have Sebulba versions of everything.